cxx/distributions/nonconjugate.hh

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <map>
+#include <random>
+#include "distributions/base.hh"
+
+template <typename T>
+class NonconjugateDistribution : public Distribution<T> {
+ public:
+  // Abstract base class for Distributions that don't have conjugate priors.
+  std::map<T, int> seen;
+
+  // The log probability of x given the current latent values.
+  virtual double logp(const T& x) const = 0;
+
+  // Sample a value from the distribution given the current latent values.
+  virtual T sample(std::mt19937* prng) = 0;
+
+  // Transition hyperparameters given the current latent values.
+  virtual void transition_hyperparameters(std::mt19937* prng) = 0;
+
+  // Set the current latent values to a sample from the parameter prior.
+  virtual void init_theta(std::mt19937* prng) = 0;
+
+  // Transition the current latent values.
+  virtual void transition_theta(std::mt19937* prng) = 0;
+
+  void incorporate(const T& x) {
```

Comment:
In general we'll want to override incorporate/unincorporate to maintain sufficient statistics and avoid the loop at each call to logp_score, right? Is that not possible for Skellam, or is this more efficient?

For distributions with mostly unique samples, keeping a running tally of logp_score at each call to incorporate/unincorporate could be better as it would avoid looping over all the data in logp_score. Ultimately I think we might want to leave incorporate/unincorporate/logp_score undefined in the base class and have each distribution explicitly do what's best, but I think this is fine for now.

Reply:
I've been told in places like https://www.johndcook.com/CompendiumOfConjugatePriors.pdf (bottom of page 2) that having fixed dimensional sufficient statistics implies the existence of a conjugate prior family.  That said, I've never actually read the precise statement or proof of that claim.

I'm fine with a running tally, though.  I'll upgrade the code to do that now.